### PR TITLE
WIP: new error handling

### PR DESCRIPTION
**Pull request by mchernov-intel**

### CODE REVIEW

**Comment by mchernov-intel** [comment]
@michael-smirnov please provide initial feedback.

File: cpp/onedal/backend/error_convertors.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*******************************************************************************
```
Comment (michael-smirnov) [inline cpp/onedal/backend/error_convertors.hpp]: As we talked offline, this file should be moved inside interop layer

File: cpp/onedal/backend/interop/common.hpp [hunk 2]
```diff
@@ -52,7 +53,10 @@ inline auto call_daal_kernel(const context_cpu& ctx, Args&&... args) {
     return dal::backend::dispatch_by_cpu(ctx, [&](auto cpu) {
         constexpr daal::CpuType daal_cpu_type = daal::CpuType::avx;
         // constexpr daal::CpuType daal_cpu_type = get_daal_cpu_type(cpu);
-        return CpuKernel<Float, daal_cpu_type>().compute(args...);
+        auto status = CpuKernel<Float, daal_cpu_type>().compute(args...);
+        if (!status) {
+            dal::backend::convert_status_to_exception(status);
+        }
```
Comment (michael-smirnov) [inline cpp/onedal/backend/interop/common.hpp]: It is a good thought to localize error conversion in one place - but I'm afraid this does not work. Do not we need to make a specific conversion for different algorithms and situations? For example, when PCA fails with some error code, it needs to be converted to the first exception, but when LogReg fails with the same code - another exception shall be raised.
For me, it is not obvious that interop call of `call_daal_kernel()` will throw an exception. From my perspective, it shall return 'Status', which can be converted after calculation (inside [backend implementation](https://github.com/mchernov-intel/daal/blob/error-handling/cpp/onedal/decomposition/pca/backend/cpu/train_kernel_cov.cpp), for example)

File: cpp/onedal/error_handling.hpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
+/*******************************************************************************
```
Comment (michael-smirnov) [inline cpp/onedal/error_handling.hpp]: Need to think better about the name of the file. It does not contain "error handling" routines, just exception classes.
I think we should be aligned with C++ here and name it "exception.hpp", or event decompose it on many files like in C++ implementations.

File: cpp/onedal/error_handling.hpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace dal {
+
+class exception{
+public:
+    virtual const char* what() const noexcept = 0;
+};
+
+class logic_error : public exception {};
+class runtime_error : public exception {};
```
Comment (michael-smirnov) [inline cpp/onedal/error_handling.hpp]: Why do you want to make these classes pure virtual?

File: cpp/onedal/error_handling.hpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+class invalid_argument :  public logic_error, public std::invalid_argument {
+public:
+    invalid_argument(const invalid_argument& other) noexcept : invalid_argument(other.what()) {}
+    invalid_argument(const char* what_arg) : std::invalid_argument(what_arg) {}
```
Comment (michael-smirnov) [inline cpp/onedal/error_handling.hpp]: Don't we want to have a constructor from `std::string` as in `std::invalid_argument`?

**Review by michael-smirnov — COMMENTED** [review]
Initial feedback

**Comment by mchernov-intel** [comment]
@michael-smirnov please, review

File: cpp/onedal/error_handling.hpp (hunk 4 above)
Comment (mchernov-intel) [reply cpp/onedal/error_handling.hpp]: both are abstract classes as we discussed on WG meetings.

File: cpp/onedal/error_handling.hpp (hunk 5 above)
Comment (mchernov-intel) [reply cpp/onedal/error_handling.hpp]: It entails dependency to std::string and haven't been discussed yet. @rlnx please, comment.

File: cpp/daal/include/services/error_handling.h [hunk 6]
```diff
@@ -32,6 +32,7 @@
 #include "services/error_indexes.h"
 #include "services/error_id.h"
 #include "services/collection.h"
+#include "services/internal/error_handling_w_a.h"
```
Comment (michael-smirnov) [inline cpp/daal/include/services/error_handling.h:35]: `_w_a` suffix tells me nothing. I suggest to think about better name

File: cpp/daal/include/services/internal/error_handling_w_a.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace interface1
+{
+    class Status;
+} // namespace interface1
+
+namespace internal
+{
+    ErrorID get_error_id(interface1::Status &s);
+} // namespace internal
```
Comment (michael-smirnov) [inline cpp/daal/include/services/internal/error_handling_w_a.h:34]: Probably it is better to move function declaration and entire `internal` namespace inside of `namespace interface1`? In that case, you don't have to specify namespaces for function parameter in this file and in error_handling.h

File: cpp/daal/src/services/error_handling.cpp [hunk 8]
```diff
@@ -610,6 +610,17 @@ services::ErrorCollectionPtr Status::getCollection() const
     return services::ErrorCollectionPtr(new services::ErrorCollection());
 }
 
+namespace internal
+{
+
+ErrorID get_error_id(Status &s) {
```
Comment (michael-smirnov) [inline cpp/daal/src/services/error_handling.cpp:616]: It is not enough to pass `const Status&`?

File: cpp/onedal/backend/interop/common_error_convertor.hpp [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+#include "daal/include/services/error_handling.h"
+#include "daal/include/services/internal/error_handling_w_a.h"
+
+namespace dal::backend::interop {
+
+#define CONVERT(error, exception, description) \
+case error: \
+    throw exception(description);
+
+void common_status_to_exception(daal::services::Status s) {
```
Comment (michael-smirnov) [inline cpp/onedal/backend/interop/common_error_convertor.hpp:28]: `Status&` at least

File: cpp/onedal/backend/interop/common_error_convertor.hpp [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    using namespace daal::services;
+    const ErrorID error = daal::services::internal::get_error_id(s);
```
Comment (michael-smirnov) [inline cpp/onedal/backend/interop/common_error_convertor.hpp:30]: `using namespace daal::services`, but `daal::services::internal::get_error_id`

File: cpp/onedal/backend/interop/common_error_convertor.hpp [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    const char * description = s.getDescription();
+
+    switch (error) {
+        CONVERT(ErrorID::ErrorPCAFailedToComputeCorrelationEigenvalues, dal::internal_error, description)
```
Comment (michael-smirnov) [inline cpp/onedal/backend/interop/common_error_convertor.hpp:34]: Why you need to have this define?

File: cpp/onedal/backend/interop/common_error_convertor.hpp [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+namespace dal::backend::interop {
+
+#define CONVERT(error, exception, description) \
+case error: \
+    throw exception(description);
+
+void common_status_to_exception(daal::services::Status s) {
+    using namespace daal::services;
+    const ErrorID error = daal::services::internal::get_error_id(s);
+    const char * description = s.getDescription();
```
Comment (michael-smirnov) [inline cpp/onedal/backend/interop/common_error_convertor.hpp:31]: Don't we want to revise descriptions of errors? Are they full enough? @rlnx are we required not to change these descriptions in future releases for some sort of compatibility?

File: cpp/onedal/decomposition/pca/backend/cpu/local_error_convertor.hpp [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+#include "onedal/backend/interop/common_error_convertor.hpp"
+
+namespace dal {
+namespace decomposition {
+namespace pca {
+namespace backend {
+
+static void local_status_to_exception(daal::services::Status s) {
+    dal::backend::interop::common_status_to_exception(s);
+}
```
Comment (michael-smirnov) [inline cpp/onedal/decomposition/pca/backend/cpu/local_error_convertor.hpp:28]: I would suggest not to divide error handling to "local" and "common" part like this, but to pass a lambda function in the `common_status_to_exception()`, which will play a role of "local" conversion routine. Is it possible?

File: cpp/onedal/exception.hpp [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+class internal_error :  public runtime_error, public std::runtime_error {
+public:
+    internal_error(const internal_error& other) noexcept : internal_error(other.what()) {}
+    internal_error(const char* what_arg) : std::runtime_error(what_arg) {}
+    const char* what() const noexcept;
+};
+
+class bad_alloc : public exception, public std::bad_alloc {
+protected:
+std::unique_ptr<char[]> str;
```
Comment (michael-smirnov) [inline cpp/onedal/exception.hpp:94]: unused

File: cpp/onedal/error_handling.hpp (hunk 4 above)
Comment (michael-smirnov) [reply cpp/onedal/error_handling.hpp]: Can you please clarify why this decision was made?

**Comment by michael-smirnov** [comment]
I suggest adding @rlnx to reviewers as well

**Comment by michael-smirnov** [comment]
Please also clarify why the decision to design base exception classes as pure virtual was made.

**Comment by rlnx** [comment]
@mchernov-intel, I suggest to move this PR to the mainline repo, once all important comments will be fixed here of course

**Comment by rlnx** [comment]
@mchernov-intel, any progress with that?

**Comment by mchernov-intel** [comment]
moved to: oneapi-src/oneDAL#660

**Comment by mchernov-intel** [comment]
@rlnx The most comments are already fixed in https://github.com/oneapi-src/oneDAL/pull/660/.

**Pull request closed without merging**